### PR TITLE
vsr: new state sync protocol 

### PR DESCRIPTION
**Pull request by matklad**
This commit implements the new state sync protocol, which fixes a couple
of long-standing liveness issues. As an unintended bonus, the new
implementation ends up being simpler, at least if measured in lines of
code!

 ## Old Protocol Problems

The original thinking was to implement state-sync as a separate
protocol, mostly orthogonal to the rest of VSR. VSR maintains the log,
state sync maintains the checkpoint. Checkpoints are committed, so
there's no doubt regarding _what_ to sync, so a separate protocol sounds
easy enough, right?

Well, turns out its' not that simple, as the two issues were discovered:

_First_, the log and the checkpoint might become physically disjoint
(checkpoint ahead of the log). The implementation fixed this case by
maneuvering the replica into recovering_head, but that sadly created a
liveness issue --- recovering_head replicas don't participate in
consensus, so it is possible for cluster to get stuck (with most of the
cluster in recovering_head) even without any storage faults!

_Second_, the log and the checkpoint might become logically inconsistent
(the message in the log for the checkpoint's slot might be from the
different view). Although checkpoint is committed, that's not true about
the log! This issue manifested itself as assertion failures in
`valid_hash_chain`. There were several band-aid fixes for the failures,
which boiled down to maneuvering replica into recovering_head if
checkpoint's view and log view are different. But we weren't able to
plug all the wholes, as it slowly became apparent that fixing everything
would require duplicating VSR-specfic SV handling logic when dealing
with checkpoints.

 ## The New Protocol

The idea is rather simple --- add primary's checkpoint to SV and handle
state sync in `on_start_view`. This makes it so that the checkpoint and
log are consistent by construction. This also yields an unexpected
benefit: a replica no longer ignores SV's which are too far in the
future, it is always possible to install a header from SV (although you
might need to update the checkpoint as well).

The implementation is somewhat tricky --- we update checkpoint
asynchronously, but we update replica state immediately, so there's a
window while our `op` is consistent with in-memory checkpoint, but not
on-disc checkpoint. This signalled by

    self.syncing == .updating_superblock

state. While in this state, the replica mostly just waits for the
superblock update to finish and doesn't do much else.

Additionally, the implementation right now is optimized for smaller
diff, rather than for path dependent perfection. For example, we still
call it `Superblock.view_change`, even though now this function can be
called _only_ to advance the checkpoint state. I want to do a follow up
cleanup PR. There's perhaps more code/asserts that needs to be adjusted
now that `.recovering_head` is again possible only after `.recovering`.

The equivalent code have been running in VOPR for quite some time, and I
think any sync-consistency related issues are fixed. The one current
outstading VOPR bug (with several seeds) is an orthogonal issue where we
allow to sync to a checkpoint before it's sync_content is done.

This particular commit has some extra assertions and cleanups, so they
might still blow up, but hopefully would be trivial to fix.

### CODE REVIEW

File: src/constants.zig
Comment (matklad) [inline src/constants.zig:326]: I randomly stuffed `vsr.CheckpointState` at the end, after the headers. Not sure where it should go at the front or at the back

File: src/vsr/replica.zig
Comment (matklad) [inline src/vsr/replica.zig:527]: `repair_sync_timeout` now directly sends the requesting message (which is just request_start_view)

File: src/vsr/replica.zig
Comment (matklad) [inline src/vsr/replica.zig:776]: transition_to_recovering_head now only happens from `.recovering` right after open

File: src/vsr/replica.zig
Comment (matklad) [inline src/vsr/replica.zig:1418]: Left the actual messages for now. I must say I am unsure how this whole things works with uprgades =/

File: src/vsr/replica.zig
Comment (matklad) [inline src/vsr/replica.zig:1601]: `self.syncing == .updating_superblock` is a weird state where we can't do anything useful, so I added a bunch of `if (`self.syncing == .updating_superblock`) return` where the vopr told me to!

File: src/vsr/replica.zig
Comment (matklad) [inline src/vsr/replica.zig:2371]: still need to come with a better name for `view_durable_update` as it no longer necessary updates the view. But I leave this for a follow up

File: src/vsr/replica.zig
Comment (matklad) [inline src/vsr/replica.zig:2316]: NB: now use the `_sync` variant!

File: src/vsr/replica.zig
Comment (matklad) [inline src/vsr/replica.zig:3163]: Unlike the old sync flow, we now don't preserve the state which tells us "we were stuck when the timeout fired" anymore. Rather, the state is preseved implictly by the network, when we send request start view, and then, when SV arrives to us, we just go to state sync.

In other words, it is possible that the priary decides to send and SV to us out of the blue, and, if we receive it and _at that point_ `repair_stuck` returns true, we'll go into the state sync, even though we weren't a lot of time in the repair_stuck state.

But I don't think we are sending SVs at random, and, if we receive one, it is _most_ likely due to us sending a request beforehand

File: src/vsr/replica.zig
Comment (matklad) [inline src/vsr/replica.zig:7995]: There might be several updates queued, so we check that the correct one finished

File: src/vsr/replica_test.zig
Comment (matklad) [inline src/vsr/replica_test.zig:1000]: We no longer have this bump sync target case at all, updates are _mostly_ atomic.

File: src/vsr/replica_test.zig
Comment (matklad) [inline src/vsr/replica_test.zig:1081]: This is the test that motivated this whole ordeal!

File: src/vsr/replica_test.zig
Comment (matklad) [inline src/vsr/replica_test.zig:2091]: just to get nicer error message when a test fails

**Comment by matklad** [comment]
The level of VOPR scrutiny this survives: 

image

It _will_ fail later with a mostly unrelated bug with shallowly-durable checkpoints, but perhaps the sync+view change parts actually work :sweat_smile:

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig]:
```suggestion
                // State sync: at this point, we know we want to replace our checkpoint
                // with the one from this SV.
```

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig]:
```suggestion
                    // Otherwise, cancel in progress commit and prepare to sync.
```

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig]:
```suggestion
                            \\{}: on_start_view: sync {s} view={} op_checkpoint={} op_checkpoint_new={}
```

Nit, but we use this format pretty much everywhere. (I use it for syntax highlighting vopr logs!)

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig]:
```suggestion
                    \\{}: on_start_view: sync started view={} op_checkpoint={} op_checkpoint_new={}
```

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig]:
```suggestion
                // The new checkpoint will we written to the superblock asynchronously.
                // From this point on, we are in a delicate state where we must be using this
                // in-memory checkpoint to check validity of log messages.
```

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig:2309]: "Replace our log with the suffix from SV." is a pun, sort of: 

- You can interpret the "suffix" as the suffix of the primary's log, i.e. the SV headers, or
- you can interpret the "suffix" as the suffix of the SV headers themselves, after we discard whichever leading headers are greater than `op_prepare_max_sync()`. :smile:

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig]:
```suggestion
                                // `vsr_state.checkpoint.commit_min` itself does not need to be synced.
```

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig]: Can this be removed?

File: src/testing/cluster.zig
Comment (sentientwaffle) [inline src/testing/cluster.zig:800]: You can remove `sync_commenced` from `log_replica`'s enum. And update `docs/about/internals/testing.md` accordingly.

And if it isn't in 2 parts, maybe we should change the symbol.

File: src/vsr/replica_test.zig
Comment (sentientwaffle) [inline src/vsr/replica_test.zig]: Before this block,

```
    try expectEqual(p.role(), .primary);
```

And after:

```
    try expectEqual(p.role(), .backup);
```

File: src/vsr/replica_test.zig
Comment (sentientwaffle) [inline src/vsr/replica_test.zig]: :thinking: Why are these changes necessary?

File: src/vsr/replica_test.zig
Comment (sentientwaffle) [inline src/vsr/replica_test.zig:1076]: Prior to this, we could add

```
try c.request(checkpoint_2_trigger + 1, checkpoint_2_trigger);
```

to demonstrate that the cluster is blocked.

File: src/vsr/replica_test.zig
Comment (sentientwaffle) [inline src/vsr/replica_test.zig:1073]: We can also test `b2.status() == .view_change`

File: src/vsr/replica_test.zig
Comment (sentientwaffle) [inline src/vsr/replica_test.zig:1176]: Without the asserts before/after b1 is restarted, its not clear to me why we restart it -- maybe we can remove the restart?

File: src/vsr/replica.zig
Comment (sentientwaffle) [reply src/vsr/replica.zig:2371]: Likewise `superblock.view_change`. Maybe combine `superblock.view_change()` and `superblock.checkpoint()` into `superblock.update(union(enum) { checkpoint, view_change, sync })`?

File: src/vsr/superblock.zig
Comment (sentientwaffle) [inline src/vsr/superblock.zig:921]: This function's comment is out of date.

File: src/vsr/sync.zig
Comment (sentientwaffle) [inline src/vsr/sync.zig]:
```suggestion
    /// requested cancellation of the commit process and entered `awaiting_checkpoint` to get
```

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig:8987]: Other preconditions:
- `commit_stage == .idle` (likewise for `sync_superblock_update_finish`)
- Sanity-check `self.syncing.updating_superblock`'s checkpoint op against our current checkpoint op.
- `maybe(self.view_durable_updating());`

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig]: In `on_repair`, add `assert(self.syncing != .updating_superblock);`.
(The ordering of the checks in `on_prepare` is important and subtle -- we check `self.syncing == .updating_superblock` before checking `is_repair`.)

File: src/constants.zig
Comment (matklad) [reply src/constants.zig:326]: Thinking more about this, stuffing fixed-sized checkpoint after the variable number of headers is just weird, so I swapped the order.

File: src/vsr/replica.zig
Comment (matklad) [inline src/vsr/replica.zig:10366]: As a follow up, looks like our header parsign fucntions are no longer polymorphic, so we could perhaps could have `Message.StartView.checkpoint`, `Message.Headers.headers`, etc.

File: src/constants.zig
Comment (sentientwaffle) [reply src/constants.zig:326]: :+1: I was going to suggest the same thing! For that reason, and also b/c that way it matches the order in the superblock header as well (where `CheckpointState` also precedes the view headers).

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig:9027]: Additionally, we could assert

```
            assert(self.superblock.staging.vsr_state.checkpoint.header.checksum ==
                stage.checkpoint_state.header.checksum);
```

to verify that even if there is another `view_durable_update` running, it isn't a state-sync update.

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig:2243]: :thinking: Shouldn't this be
```suggestion
                message.header.commit > self.op_prepare_max() + constants.journal_slot_count or
```
instead?

File: src/vsr/replica.zig
Comment (matklad) [reply src/vsr/replica.zig:1418]: (got the plan here now, would add some more commits to implement it)

File: src/vsr/replica.zig
Comment (matklad) [reply src/vsr/replica.zig:2371]: Maybe even something like

```
fn update(view: ?ViewUpdate, checkpont: ?CheckpointUpdate) {
    assert(!(view == null and checkpoint == null));
}

```

File: src/vsr/replica.zig
Comment (matklad) [reply src/vsr/replica.zig]: That's not a nit at all, that's really important to make our log messages consistent, thanks!

File: src/vsr/replica.zig
Comment (matklad) [reply src/vsr/replica.zig]: Good catch! At some point I should go over the original sync PR and see if there are more recovoinrg_head logic which can be removed, or asserts tightened.

File: src/testing/cluster.zig
Comment (matklad) [reply src/testing/cluster.zig:800]: Removed the enum, but I'll leave further adjustments for later. It feels like maybe the entire `sync_dispatch` machinery no longer pulls its weight.

File: src/vsr/replica_test.zig
Comment (matklad) [reply src/vsr/replica_test.zig]: Good catch, they aren't, backed them out! 

As why I have I added them, I sadly don't remeber. The only thing I recall about fixing this test is how the comment speaks about `A0` but we use `p` as a variable named

File: src/vsr/replica_test.zig
Comment (matklad) [reply src/vsr/replica_test.zig:1176]: Added the following comment:

```
// Wipe B1 in-memory state and check that it ends up in a consistent state after restart. 
```

What we are testing here is that b1 doesn't trip `assert(valid_hash_chain)` assertion, so there isn't anything specific to test for here.

File: src/vsr/replica.zig
Comment (matklad) [reply src/vsr/replica.zig:9027]: Oh, this is excellent!

File: src/vsr/replica.zig
Comment (matklad) [reply src/vsr/replica.zig:2243]: Good question. I think either version is correct, and, moreover, we could completely remove this condition (we'll just hit `repair_stuck` a bit later in the future).

I'll leave this as is, though, because it seems like a nice heuristic for when to switch to state-sync proactively, and because this allows to comment on the subtelty here (which the original comment missed):

The overall condition checkifs that SV's commit is ahead of our current checkpont's prepare_max. 

**If** SV's checkpoint is our `op_checkpoint_next`, then `message.header.commit > self.op_prepare_max()` guarantees that that checkpoint is durable and is safe to sync to. 

**However**, if SV's checkpoint is further in the future, it might be the case that it is not durable. Eg, we are a severely lagging replica, primary just triggered a checkpoint and tells this fresh checkpoint to us, but the rest of the cluster is still slightly behind the cluster. But in this case it is still save to sync to that checkpoint, because our own prepares are no longer relevant. 

---

I guess if we want to tighten the heuristic where to skip the state sync, we should check which prepare we are actually missing, and sync when that gets evicted from primary's log.

**Review by sentientwaffle — DISMISSED** [review]
LGTM! (Though I assume this shouldn't be merged as-is, depending on what the transition/upgrade strategy will look like.)

**Comment by matklad** [comment]
>Though I assume this shouldn't be merged as-is, depending on what the transition/upgrade strategy will look like.

Indeed! I've now added the upgrade strategy as the last commit, which I _think_ should work.

Not ready to land that right away ether though:

* I want to test in in vopr (in fact, I already have a failing seed on hand)
* The VOPR tests (in https://github.com/tigerbeetle/tigerbeetle/pull/2176) are somewhat hacky, as they make `vsr.Version` overridable at runtime. I am not fully sure if I want to land that approach (it adds a bunch of vopr-only code), but, right now, I am more curious whether the actual upgrade path will work.

**Comment by matklad** [comment]
Ok, I realized that I got myself off-by-oned here! What https://github.com/tigerbeetle/tigerbeetle/pull/2176) aims to test is the the release after the next one, when we switch to only having version 1.

I _think_ to the extent that the switch logic could be tested for the next release, it _is_ tested in this PR as is: because the replicas send both SV0 and SV1, VOPR dropping `SV1` message simulates a cluster with some old replicas. 

What is _not_ tested by the VOPR is that the old memebers can state sync (the latest commit adds back the code to respond to `fn request_sync_target`, but no one actually requests it yet). To test _that_ in VOPR, I'll need to somehow bring back the old state sync logic, and that seems not really feasible to do without adding a bunch of effectively vopr-only code

File: src/vsr.zig
Comment (sentientwaffle) [inline src/vsr.zig]: The comment is out of date.

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig]:
```suggestion
            current: *Message.StartViewDeprecated,
```

(I know they are technically identical.)

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig:4693]: If we are using `.get_message(.start_view)`, could we comptime-assert `Header.StartView == Header.StartViewDeprecated` here?

**Comment by sentientwaffle** [comment]
Other than those nits :arrow_up: this LGTM.

**Comment by matklad** [comment]
Going to leave this waiting until we make a release with https://github.com/tigerbeetle/tigerbeetle/pull/2181. After that, fix nits, rebase on main, and should be good to go!

**Review by sentientwaffle — APPROVED** [review]
LGTM to merge after the next release!

**Pull request merged**
